### ISSUE TITLE
Implement IteratorClose

### ISSUE THREAD

**Issue by wolfbeast** [MoonchildProductions/UXP#74]
Blocks #71 

[Bug 1147371](https://bugzilla.mozilla.org/show_bug.cgi?id=1147371) - Implement IteratorClose

This depends on at least #73

**Comment by janekptacijarabaci** [comment]
__Some test/personal branch:__
[tree@js_IteratorClose_1](https://github.com/janekptacijarabaci/UXP/tree/js_IteratorClose_1)

I've created the new build (x32, Windows) and tested (shortly) but it's not done yet. There are a lot of other bugs (in progress...).

**Comment by janekptacijarabaci** [comment]
__Some test/personal branch:__
[tree@js_IteratorClose_1](https://github.com/janekptacijarabaci/UXP/tree/js_IteratorClose_1)
(I've created the new build (x32, Windows) and tested)

``` javascript
function assertEq(a, b) {
  console.log(a.toString() + " == " + b + ": " + (a == b));
}

var a = (function () {
    for (var x in [0]) {
        try {} finally {
            return 11;
        }
    }
})();
assertEq(a, 11);

var b = (function () {
    for (var x of [0]) {
        try {} finally {
            return 12;
        }
    }
})();
assertEq(b, 12);

var c = (function () {
    for (var x in [0]) {
        for (var y of [0]) {
            try {} finally {
                return 13;
            }
        }
    }
})();
assertEq(c, 13);

var d = (function () {
    for (var x in [0]) {
        for (var y of [0]) {
            try {} finally {
                for (var z in [0]) {
                    for (var w of [0]) {
                        try {} finally {
                            return 14;
                        }
                    }
                }
            }
        }
    }
})();
assertEq(d, 14);

/*
After implemented bug 1147371:

[crash]

Expected results:

11 == 11: true
12 == 12: true
13 == 13: true
14 == 14: true
*/
```
``` javascript
function assertEq(a, b) {
  console.log(a.toString() + " == " + b + ": " + (a == b));
}

var called = 0;
function reset() {
  called = 0;
}
var obj = {
  [Symbol.iterator]() {
    return {
      next() {
        return { value: 10, done: false };
      },
      return() {
        called++;
        return {};
      }
    };
  }
};

var a = (function () {
    for (var x in [0]) {
        try {} finally {
            return 11;
        }
    }
})();
assertEq(a, 11);

reset();
var b = (function () {
    for (var x of obj) {
        try {} finally {
            return 12;
        }
    }
})();
assertEq(called, 1);
assertEq(b, 12);

reset();
var c = (function () {
    for (var x in [0]) {
        for (var y of obj) {
            try {} finally {
                return 13;
            }
        }
    }
})();
assertEq(called, 1);
assertEq(c, 13);

reset();
var d = (function () {
    for (var x in [0]) {
        for (var y of obj) {
            try {} finally {
                for (var z in [0]) {
                    for (var w of obj) {
                        try {} finally {
                            return 14;
                        }
                    }
                }
            }
        }
    }
})();
assertEq(called, 2);
assertEq(d, 14);

/*
After implemented bug 1147371:

[crash]

Expected results:

11 == 11: true
1 == 1: true
12 == 12: true
1 == 1: true
13 == 13: true
2 == 2: true
14 == 14: true
*/
```
``` javascript
var iterable = {
    [Symbol.iterator]() {
        return {
            next(x) {
                console.log("NEXT called");
                return { done: false };
            },
            return(x) {
                console.log("RETURN called");
                return { done: true };
            },
            throw(e) {
                console.log("THROW called");
                throw e;
            },
        };
    }
};

function* y() {
    yield* iterable;
}

var g1 = y();
g1.next();
try { g1.throw(); } catch (e) { }

/*
After implemented bug 1147371:

NEXT called
THROW called
RETURN called

Expected results:

NEXT called
THROW called
*/
```
``` javascript
function assertEq(a, b) {
  console.log(a.toString() + " == " + b + ": " + (a == b));
}

function assertThrowsValue(a, b) {
  try {
    a();
  } catch (e) {
    console.log(e + " == " + b + ": " + (e == b));
  }
}

function throwlhs() {
  throw "in lhs";
}

function makeIterator(overrides) {
  var iterator = {
    next: function() {
      if (overrides && overrides.next)
        return overrides.next();
      return { done: false };
    },
    return: function() {
      if (overrides && overrides.ret)
        return overrides.ret();
      return { done: true };
    }
  };

  return function() { return iterator; };
};

    var returnCalled = 0;
    var returnCalledExpected = 0;
    var iterable = {};    
    // throw in lhs ref calls IteratorClose with falsy "done".
    iterable[Symbol.iterator] = makeIterator({
        next: function() {
            // "done" is undefined.
            return {};
        },
        ret: function() {
            returnCalled++;
            return {};
        }
    });
    assertThrowsValue(function() {
        0, [...{}[throwlhs()]] = iterable;
    }, "in lhs");
    assertEq(returnCalled, ++returnCalledExpected);

/*
Expected results:

in lhs == in lhs: true
3 == 3: true
*/
```
``` javascript
function assertEq(a, b) {
  console.log(a.toString() + " == " + b + ": " + (a == b));
}

function assertThrowsValue(a, b) {
  try {
    a();
  } catch (e) {
    console.log(e + " == " + b + ": " + (e == b));
  }
}

function makeIterator(overrides) {
  var iterator = {
    next: function() {
      if (overrides && overrides.next)
        return overrides.next();
      return { done: false };
    },
    return: function() {
      if (overrides && overrides.ret)
        return overrides.ret();
      return { done: true };
    }
  };

  return function() { return iterator; };
};

    var returnCalled = 0;
    var returnCalledExpected = 0;
    var catchEntered = 0;
    var finallyEntered = 0;
    var finallyEnteredExpected = 0;
    var iterable = {};
    iterable[Symbol.iterator] = makeIterator({
        ret: function() {
            returnCalled++;
            throw 42;
        }
    });

    // inner try cannot catch IteratorClose throwing
    assertThrowsValue(function() {
        for (var x of iterable) {
            try {
                return;
            } catch (e) {
                catchEntered++;
            } finally {
                finallyEntered++;
            }
        }
    }, 42);
    assertEq(returnCalled, ++returnCalledExpected);
    assertEq(catchEntered, 0);
    assertEq(finallyEntered, ++finallyEnteredExpected);

/*
Expected results:

42 == 42: true
1 == 1: true
0 == 0: true
1 == 1: true
*/
```
``` javascript

var iterable = {};
var iterator = {
  return: 1
};
iterable[Symbol.iterator] = function() {
  return iterator;
};
for ([ class get {} ().iterator ] of [iterable]) {}

/*
After implemented bug 1147371:

[crash]

Expected results:

TypeError: 1 is not a function
*/
```
``` javascript
function* wrap() {
  let iter = {
    [Symbol.iterator]() {
      return this;
    },
    next() {
      return { value: 10, done: false };
    },
    return() {
      console.log("return is called");
      return {};
    }
  };
  for (const i of iter) {
    yield i;
  }
}
for (const i of wrap()) {
  break;
}

/*
Expected result:

"return is called"
*/
```
``` javascript
function assertEq(a, b) {
  console.log(a.toString() + " == " + b + ": " + (a == b));
}

function assertThrowsInstanceOf(a, b) {
  try {
    a();
    console.log(a + " - " + b + ": false");
  } catch (e) {
    console.log(a + " - " + b + "?: " + e.name);
  }
}

function assertThrowsValue(a, b) {
  try {
    a();
  } catch (e) {
    console.log(e + " == " + b + ": " + (e == b));
  }
}

function makeIterator(overrides) {
  var iterator = {
    next: function() {
      if (overrides && overrides.next)
        return overrides.next();
      return { done: false };
    },
    return: function() {
      if (overrides && overrides.ret)
        return overrides.ret();
      return { done: true };
    }
  };

  return function() { return iterator; };
};

    var returnCalled = 0;
    var returnCalledExpected = 0;
    var iterable = {};
    iterable[Symbol.iterator] = makeIterator({
        ret: function() {
            returnCalled++;
            return {};
        }
    });

    // break calls iter.return
    for (var x of iterable)
        break;
    assertEq(returnCalled, ++returnCalledExpected);

    // throw in body calls iter.return
    assertThrowsValue(function() {
        for (var x of iterable)
            throw "in body";
    }, "in body");
    assertEq(returnCalled, ++returnCalledExpected);

    // throw in lhs ref calls iter.return
    function throwlhs() {
        throw "in lhs";
    }
    assertThrowsValue(function() {
        for ((throwlhs()) of iterable)
            continue;
    }, "in lhs");
    assertEq(returnCalled, ++returnCalledExpected);

    // throw in iter.return doesn't re-call iter.return
    iterable[Symbol.iterator] = makeIterator({
        ret: function() {
            returnCalled++;
            throw "in iter.return";
        }
    });
    assertThrowsValue(function() {
        for (var x of iterable)
            break;
    }, "in iter.return");
    assertEq(returnCalled, ++returnCalledExpected);

    // throw in iter.next doesn't call IteratorClose
    iterable[Symbol.iterator] = makeIterator({
        next: function() {
            throw "in next";
        }
    });
    assertThrowsValue(function() {
        for (var x of iterable)
            break;
    }, "in next");
    assertEq(returnCalled, returnCalledExpected);

    // "return" must return an Object.
    iterable[Symbol.iterator] = makeIterator({
        ret: function() {
            returnCalled++;
            return 42;
        }
    });
    assertThrowsInstanceOf(function() {
        for (var x of iterable)
            break;
    }, TypeError);
    assertEq(returnCalled, ++returnCalledExpected);

    // continue doesn't call iter.return for the loop it's continuing
    var i = 0;
    iterable[Symbol.iterator] = makeIterator({
        next: function() {
            return { done: i++ > 5 };
        },
        ret: function() {
            returnCalled++;
            return {};
        }
    });
    for (var x of iterable)
        continue;
    assertEq(returnCalled, returnCalledExpected);

    // continue does call iter.return for loops it skips
    i = 0;
    L: do {
        for (var x of iterable)
            continue L;
    } while (false);
    assertEq(returnCalled, ++returnCalledExpected);

/*
Actual results:

0 == 1: false
in body == in body: true
0 == 2: false
in lhs == in lhs: true
0 == 3: false
0 == 4: false
in next == in next: true
0 == 4: false
function () {
        for (var x of iterable)
            break;
    } - function TypeError() {
    [native code]
}: false
0 == 5: false
0 == 6: false

Expected results:

1 == 1: true
in body == in body: true
2 == 2: true
in lhs == in lhs: true
3 == 3: true
in iter.return == in iter.return: true
4 == 4: true
in next == in next: true
4 == 4: true
function () {
        for (var x of iterable)
            break;
    } - function TypeError() {
    [native code]
}?: TypeError
5 == 5: true
6 == 6: true
*/
```

![001](https://user-images.githubusercontent.com/2373486/37879219-90b9b6da-3064-11e8-84fd-e7fd13d8bf1c.png)
![002](https://user-images.githubusercontent.com/2373486/37879220-910451d6-3064-11e8-828b-924823ba0d3b.png)
![009](https://user-images.githubusercontent.com/2373486/37879221-9122b072-3064-11e8-8896-365ffba4270d.png)

**Comment by wolfbeast** [comment]
Thanks for your efforts!

Resolved with merge e19749682050ff716fc9ff3bbc05ee3911570670
